lib/oban/telemetry.ex

```diff
@@ -158,4 +158,19 @@ defmodule Oban.Telemetry do
       |> Jason.encode!()
     end)
   end
+
+  @doc """
+  Measure the time it took to run a function and publishes a telemetry
+  """
+  @spec span(name :: atom(), fun :: (() -> term()), metadata :: map()) :: term()
+  def span(name, fun, metadata \\ %{}) do
+    start_mono = System.monotonic_time(:microsecond)
+    result = fun.()
+
+    :telemetry.execute([:oban, name], %{duration: duration(start_mono)}, metadata)
+
+    result
```

Comment:
To match more closely to the telemetry "spec", we could emit multiple events:

```elixir
    start_mono = System.monotonic_time(:microsecond)
    :telemetry.execute([:oban, name, :start], %{system_time: start_mono}, metadata)

    try do
      result = fun.()
      :telemetry.execute([:oban, name, :stop], %{duration: duration(start_mono)}, metadata)
      result
    catch
      kind, reason ->
        stacktrace = System.stacktrace()

        :telemetry.execute([:oban, name, :exception], %{duration: duration(start_mono)}, Map.merge(metadata, %{kind: kind, reason: reason, stacktrace: stacktrace}))

        :erlang.raise(kind, reason, stacktrace)
    end
```

Reply:
What do you think, @sorentwo?

Reply:
FYI, I culled this example primarily from teamon/tesla#324, but there are multiple implementations linked to that beam-telemetry issue above that you can reference, too.

Reply:
The implementation looks good. I've modified it with a few changes which I'll call out below:

```elixir
  start_time = System.system_time()
  start_mono = System.monotonic_time(:microsecond)

  :telemetry.execute([:oban, name, :start], %{start_time: start_time}, metadata)

  try do
    result = fun.()

    :telemetry.execute([:oban, name, :stop], %{duration: duration(start_mono)}, metadata)

    result
  catch
    kind, reason ->
      :telemetry.execute(
        [:oban, name, :exception],
        %{duration: duration(start_mono)},
        Map.merge(metadata, %{kind: :error, error: error, stack: __STACKTRACE__})
      )

      :erlang.raise(kind, reason, __STACKTRACE__)
  end
```

1. Use `start_time` instead of `system_time` for consistency with other Oban telemetry
2. Use `__STACKTRACE__` instead because we don't support any versions prior to OTP 21
3. Use kind, error and stack for consistency with the other Oban telemetry

The inconsistency in 1 and 3 will need to be addressed, along with using native time instead of microseconds, in a future PR.

Reply:
@sorentwo on the exception, I'd highly encourage using `:stacktrace` for the key as we'll be looking for that key in the tracing handlers. If you have other places using the `:stack` convention, I would suggest duplicating to both keys for now and deprecate it. The standard convention we've outlined is `[:kind, :error, :stacktrace]`.

And yes, monotonic times are expected to be in native.